File: python3/agent.py
from math import pi
import random


class Agent:
    def __init__(self):
        self.alpha = 0.1  # 学習率
        self.gamma = 0.999  # 割引率
        self.epsilon = 0.1  # ランダムに行動選択する割合

        self.init_qvalue = 10000.0  # q-value の初期値
        self.actions = [-10.0, 10.0]  # 行動の候補

        # 状態分割の下限と上限
        self.x_limits = [-2.0, 2.0]
        self.theta_limits = [-pi, pi]
        self.xdot_limits = [-2.0, 2.0]
        self.thetadot_limits = [-10.0, 10.0]

        # 状態の分割数
        self.x_num = 4
        self.theta_num = 90
        self.xdot_num = 10
        self.thetadot_num = 50

        # 状態分割の bins を生成
        self.x_bins = self.make_bins(self.x_limits, self.x_num)
        self.theta_bins = self.make_bins(self.theta_limits, self.theta_num)
        self.xdot_bins = self.make_bins(self.xdot_limits, self.xdot_num)
        self.thetadot_bins = self.make_bins(
            self.thetadot_limits, self.thetadot_num)

        self.qtable = self.make_qtable()

    def action(self, s):
        """eps-greedy"""
        if random.random() < self.epsilon:
            return random.choice(self.actions)

        s_idx = self.get_s_idx(s)
        max_idx = self.argmax(self.qtable[s_idx])
        return self.actions[max_idx]

    def learn(self, s, a, r, snext):
        s_idx = self.get_s_idx(s)
        a_idx = self.actions.index(a)
        snext_idx = self.get_s_idx(s)

        self.qtable[s_idx][a_idx] = \
            (1.0 - self.alpha) * self.qtable[s_idx][a_idx] + \
            self.alpha * (r + self.gamma * max(self.qtable[snext_idx]))

    def set_test_params(self):
        """test 用のパラメータに変更する"""
        self.alpha = 0.0
        self.epsilon = 0.0

    def make_qtable(self):
        # NOTE: すごい多重配列にすると table 作るだけで心折れるので見送り

        total_states_num = self.x_num * self.theta_num * self.xdot_num * self.thetadot_num
        return [[self.init_qvalue] * len(self.actions) for _ in range(total_states_num)]

    def make_bins(self, limits, num):
        width = (limits[1] - limits[0]) / (num - 2)
        return [limits[0] + width * i for i in range(num - 1)]

    def digitize(self, bins, x):
        for i, v in enumerate(bins):
            if x < v:
                return i
        return len(bins)

    def digitize_all(self, s):
        x_idx = self.digitize(self.x_bins, s[0])
        theta_idx = self.digitize(self.theta_bins, s[1])
        xdot_idx = self.digitize(self.xdot_bins, s[2])
        thetadot_idx = self.digitize(self.thetadot_bins, s[3])
        return x_idx, theta_idx, xdot_idx, thetadot_idx

    def get_s_idx(self, s):
        indices = self.digitize_all(s)
        return indices[0] + self.x_num * (indices[1] + self.theta_num * (indices[2] + self.xdot_num * indices[3]))

    def argmax(self, lst):
        idx = 0
        max_val = lst[0]
        for i, v in enumerate(lst):
            if v > max_val:
                idx = i
        return idx

File: ruby/agent.rb
class Agent
  def initialize
    @alpha = 0.1  # 学習率
    @gamma = 0.999  # 割引率
    @epsilon = 0.1  # ランダムに行動選択する割合

    @init_qvalue = 10000.0  # q-value の初期値
    @actions = [-10.0, 10.0]  # 行動の候補

    # 状態分割の下限と上限
    @x_limits = [-2.0, 2.0]
    @theta_limits = [-Math::PI, Math::PI]
    @xdot_limits = [-2.0, 2.0]
    @thetadot_limits = [-10.0, 10.0]

    # 状態の分割数
    @x_num = 4
    @theta_num = 90
    @xdot_num = 10
    @thetadot_num = 50

    # 状態分割の bins を生成
    @x_bins = make_bins(@x_limits, @x_num)
    @theta_bins = make_bins(@theta_limits, @theta_num)
    @xdot_bins = make_bins(@xdot_limits, @xdot_num)
    @thetadot_bins = make_bins(@thetadot_limits, @thetadot_num)

    @qtable = make_qtable
  end

  # eps-greedy
  def action(s)
    if Random.rand < @epsilon
      return @actions.sample
    end

    s_idx = s_index(s)
    max_idx = argmax(@qtable[s_idx])
    @actions[max_idx]
  end

  def learn(s, a, r, snext)
    s_idx = s_index(s)
    a_idx = @actions.index(a)
    snext_idx = s_index(snext)

    @qtable[s_idx][a_idx] = \
      (1.0 - @alpha) * @qtable[s_idx][a_idx] + \
      @alpha * (r + @gamma * @qtable[snext_idx].max)
  end

  def set_test_params
    @alpha = 0.0
    @epsilon = 0.0
  end

  def make_bins(limits, num)
    width = (limits[1] - limits[0]) / (num - 2)
    (0..(num-2)).map { |i| limits[0] + width * i }
  end

  def make_qtable
    s_size = @x_num * @theta_num * @xdot_num * @thetadot_num
    Array.new(s_size) {
      Array.new(@actions.size, @init_qvalue)
    }
  end

  def digitize(bins, x)
    bins.each_with_index do |v, i|
      if x < v
        return i
      end
    end
    bins.size
  end

  def digitize_all(s)
    x_idx = digitize(@x_bins, s[0])
    theta_idx = digitize(@theta_bins, s[1])
    xdot_idx = digitize(@xdot_bins, s[2])
    thetadot_idx = digitize(@thetadot_bins, s[3])
    [x_idx, theta_idx, xdot_idx, thetadot_idx]
  end

  def s_index(s)
    indices = digitize_all(s)
    indices[0] + @x_num * (indices[1] + @theta_num * (indices[2] + @xdot_num * indices[3]))
  end

  def argmax(arr)
    idx = 0
    max_val = arr[0]
    arr.each_with_index do |v, i|
      if v > max_val
        idx = i
      end
    end
    idx
  end
end